File: exercises/day10-TimePortal/scala/src/main/scala/timeportal/Traveler.scala
package timeportal

class Traveler(val name: String, val visaCode: String):
  private var _currentYear: Int = 2024
  private var _jumpCount: Int = 0
  private var _healthStatus: String = "STABLE" // "STABLE", "UNSTABLE", "CRITICAL"
  private var lastJumpTimestamp: Long = 0

  def currentYear: Int = _currentYear
  def jumpCount: Int = _jumpCount

  def hasTemporalSickness: Boolean =
    if _jumpCount > 10 then
      true
    else
      if System.currentTimeMillis() - lastJumpTimestamp < 3600000 then
        true
      else
        false

  def incrementJumpCount(): Unit =
    _jumpCount += 1
    lastJumpTimestamp = System.currentTimeMillis()
    if _jumpCount > 5 then
      _healthStatus = "UNSTABLE"
    if _jumpCount > 10 then
      _healthStatus = "CRITICAL"

  def setCurrentYear(year: Int): Unit = _currentYear = year
  def getHealthStatus: String = _healthStatus

File: exercises/day10-TimePortal/csharp/TimePortal/TimePortal/Traveler.cs
namespace TimePortal;

public class Traveler
{
    private string _name;
    private string _visaCode;
    private int _currentYear;
    private int _jumpCount;
    private string _healthStatus; // "STABLE", "UNSTABLE", "CRITICAL"
    private long _lastJumpTimestamp;

    public Traveler(string name, string visaCode)
    {
        _name = name;
        _visaCode = visaCode;
        _currentYear = 2024;
        _jumpCount = 0;
        _healthStatus = "STABLE";
        _lastJumpTimestamp = 0;
    }

    public bool HasTemporalSickness()
    {
        if (_jumpCount > 10)
        {
            return true;
        }
        else
        {
            if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - _lastJumpTimestamp < 3600000)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
    }

    public void IncrementJumpCount()
    {
        _jumpCount++;
        _lastJumpTimestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (_jumpCount > 5)
        {
            _healthStatus = "UNSTABLE";
        }
        if (_jumpCount > 10)
        {
            _healthStatus = "CRITICAL";
        }
    }

    public string Name => _name;
    public string VisaCode => _visaCode;
    public int CurrentYear => _currentYear;
    public void SetCurrentYear(int year) => _currentYear = year;
    public string GetHealthStatus() => _healthStatus;
    public int JumpCount => _jumpCount;
}
